Dev/src gen cell value mapper

Ashymonth opened the pull request:
Preview version of source gen with ```CellValueMapper```

Ashymonth commented:
This is how i think to implement source gen with ```CellValueMapper```. 
I'd like to discus a few questions before i start to implement full functional. 

1. Should i take care of ```CellValueTruncateAttribute```? I think no but maybe i missed smth. Should i raise warning message  if user applied ```CellValueTruncateAttribute``` and ```CellValueMapperAttribute```
2. I made array of ```ICellValueMapper``` static. It's correct?
3. I'd like to add ability to use ```CellValueMapperAttribute``` on class and property. In my case i have 50+ properties on class and 1 behavior to handle null value in properties. And if attribute exist on class and on property the type provided in property will be used. Do you agree with this feature?.

sveinungf commented:
Thanks for the PR!

I made a quick suggestion how to support this in #67, but after some more thought I'd like to make some changes to my initial suggestion:

- Instead of `CellValueMapper`, I think `CellValueConverter` is a better name. Both EF Core and `System.Text.Json` use the term "converter" for similar functionality.
- I think we should use an abstract class instead of an interface. It will then be possible to add more functionality to `CellValueConverter`s later without introducing a breaking change. The dotnet team also uses abstract classes for similar cases: https://github.com/dotnet/runtime/discussions/84120
- I would like `CellValueConverter`s to only handle cell values, and not styling, formulas, etc. It will then be more modular and can be used together with other source generator attributes that will be added in the future. This basically means the abstract method on `CellValueConverter` should have return type `DataCell`.
- We need to add a constructor to `StyledCell` with signature `StyledCell(DataCell dataCell, StyleId? styleId)` so that the source generator can create `StyledCell` when using a `CellValueConverter`.

Here is how that would look in code:
```csharp
public class MyObject
{
    [CellValueConverter(typeof(NullToDashConverter))]
    public int? Number { get; set; }
}

public class NullToDashConverter : CellValueConverter<int?>
{
    public DataCell ConvertToCell(int? value)
    {
        return value is null ? new DataCell("-") : new DataCell(value.Value);
    }
}

public abstract class CellValueConverter<T>
{
    public abstract DataCell ConvertToCell(T value);
}
```

Support for selecting a style based on the cell value probably doesn't belong in this PR, but to handle that I am considering a similar approach. That feature would depend on getting a style by its name. In code it could look like this:
```csharp
public class MyObject
{
    // TODO: Better name for the attribute?
    [CellStyleProvider(typeof(NullableIntStyleProvider))]
    public int? Number { get; set; }
}

public class NullableIntStyleProvider : CellStyleProvider<int?>
{
    public string? GetStyleName(int? value)
    {
        return value is null ? null : "My style";
    }
}

public abstract class CellStyleProvider<T>
{
    public abstract string? GetStyleName(T value);
}
```

Now, to answer your questions:
1. `CellValueConverter` will be incompatible with `CellValueTruncate`, so we should emit an error if attempting to use both attributes on a property.
2. I think it can be a static array. We should avoid duplicates in that array, so if two properties use the same `CellValueConverter`, it should only lead to a single instance in the array. Note that it is also possible to use multiple `WorksheetRow` attributes on a single context class, which will lead to multiple `RowType`s in the source generator. For that case we will need to avoid duplicate static arrays in the generated code.
3. It makes sense to be able to set a `CellValueConverter` on a higher level, either on the class or perhaps even on the whole worksheet/spreadsheet. Using the same attribute on the class level as is being used on properties could work, I'm just a bit unsure about the readability, if it becomes unclear what properties the converter applies to. I'm not sure what the best approach here would be.

Ashymonth commented:
I decided not to add support for the class level attribute for this time, because it requires another attribute with a different name and the ability to add it multiple times for different types. The current ```CellValueConverterAttribute``` cannot be applied multiple times to a property. I also added support for combining ```CellValueConverterAttribute``` and ```CellStyleAttribute```. 
I also added new analyzer errors, but I didn't include them in the ```AnalyzerReleases.Shipped.md``` file because I don't know the release policy.

sveinungf commented:
Thanks! 👍 I'll have a look.
Regarding having an attribute on class level, that can be added later. I would like to have some kind of common approach of handling defaults, but I haven't figured out how that should look like yet. The new analyzer errors I will add to that file when I release the next version.

sveinungf commented on SpreadCheetah/SourceGeneration/CellValueConverterAttribute.cs:
```diff
@@ -0,0 +1,13 @@
+/// Instructs the SpreadCheetah source generator to use specified cell mapper to create a cell.
+/// </summary>
+[AttributeUsage(AttributeTargets.Property, AllowMultiple = false, Inherited = true)]
+public sealed class CellValueConverterAttribute : Attribute
```
Maybe we can set the `Type` as a parameter in a primary constructor? E.g.
```csharp
public sealed class CellValueConverterAttribute(Type converterType) : Attribute;
```
It will enforce that a type is actually set, since the attribute can't really be used without it. It will also lead to a bit less verbose code when using the attribute.

For `InheritColumnsAttribute` it made sense to use a property, since that one had an optional parameter.

Ashymonth replied:
of course 
I forgot about this

sveinungf commented on SpreadCheetah.SourceGenerator/Extensions/AttributeDataExtensions.cs:
```diff
@@ -10,6 +10,8 @@ namespace SpreadCheetah.SourceGenerator.Extensions;
 
 internal static class AttributeDataExtensions
 {
+    private const string CellValueMapperInterfaceName = "ICellValueMapper";
```
This is not correct anymore I think.

Ashymonth replied:
done

sveinungf commented on SpreadCheetah.SourceGenerator/WorksheetRowGenerator.cs:
```diff
@@ -383,6 +391,42 @@ private static Dictionary<CellStyle, int> GenerateCreateWorksheetRowDependencyIn
+        }
+        
+        sb.AppendLine(FormattableString.Invariant($$"""
+                               private static Dictionary<string, object> _cellValueConverters = new Dictionary<string, object>
```
This seems to work correctly, but it has the disadvantage of a dictionary lookup per cell, and a cast from object to get the correct type. This can potentially affect performance in a negative way.

Perhaps we should instead consider using a static field per converter type. For example:
```csharp
private static readonly StringValueConverter _valueConverter0 = new StringValueConverter();
private static readonly NullableIntValueConverter _valueConverter1 = new NullableIntValueConverter();
private static readonly DecimalValueConverter _valueConverter2 = new DecimalValueConverter();
```
Using name `_valueConverter` plus a number can avoid potential naming collisions. The source generator would need to keep track of which field corresponds to which converter. This can be done with a Dictionary, similar to what `GenerateCreateWorksheetRowDependencyInfo` is doing.

Ashymonth replied:
yes, you're right  thanks

sveinungf commented on SpreadCheetah.SourceGenerator.SnapshotTest/Tests/WorksheetRowGeneratorCellValueConverterTests.cs:
```diff
@@ -0,0 +1,128 @@
+
+namespace SpreadCheetah.SourceGenerator.SnapshotTest.Tests;
+
+public class WorksheetRowGeneratorCellValueConverterTests
```
First of all, it's awesome that you have written tests as well 👍
These follow the same style as some of the older tests I've written. Going forward I want to have these kind of tests more self-contained, meaning that the whole code is part of each test, and that they don't reference test classes from elsewhere. The tests in `CellStyleTests.cs` are an example of this.

Ashymonth replied:
done

sveinungf commented:
This looks very good, thank you! I will make some minor adjustments and once that is ready I will release a new version on NuGet. I hope to have it ready during the weekend.

Ashymonth commented:
okay 
thank you for the help and awesome library)
